### DISCUSSION TITLE
Docker 核心技术与实现原理 · /docker

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#516] [Announcements]
https://draveness.me/docker

**Comment by Durant35** [comment]
赞一个！

**Comment by Jenny-O** [comment]
666，感谢分享干货，已推荐到 SegmentFault 头条，欢迎来 SegmentFault 分享你的技术沉淀 (๑•̀ㅂ•́)و✧
链接如下：https://segmentfault.com/p/1210000012239300

**Comment by wujunze** [comment]
很棒

**Comment by ianwoolf** [comment]
赞

**Comment by liverpoolpjy** [comment]
>当 Docker 服务器在主机上其中之后会创建新的虚拟网桥 docker0，随后在该主机上启动的全部服务在某人情况下都与该网桥相连。

有点不明白这里说的是什么。

**Comment by draveness** [comment]
> @liverpoolpjy
> >当 Docker 服务器在主机上其中之后会创建新的虚拟网桥 docker0，随后在该主机上启动的全部服务在某人情况下都与该网桥相连。
> 
> 有点不明白这里说的是什么。

其中 -> 启动, 某人 -> 默认 😓

**Comment by iScript** [comment]
图使用什么工具画的

**Comment by snakeliwei** [comment]
图使用什么工具画的+1

**Comment by draveness** [comment]
> 图使用什么工具画的

@iScript @snakeliwei 看一下文章结尾吧

**Comment by snakeliwei** [comment]
> @Draveness
> > 图使用什么工具画的
> 
> @iScript @snakeliwei 看一下文章结尾吧

Sketch可以保存为矢量么

**Comment by ruidanwang** [comment]
image 和 container的关系图是不是反了啊？

**Comment by draveness** [comment]
> @ruidanwang
> image 和 container的关系图是不是反了啊？

还真是，我回头改一下吧

**Comment by llitfkitfk** [comment]
赞

**Comment by letian0805** [comment]
纠正下文章的开头，是“容器技术”，不是“虚拟化技术”

**Comment by draveness** [comment]
> @letian0805
> 纠正下文章的开头，是“容器技术”，不是“虚拟化技术”

有权威的出处么，我没有找到可信的资料，比如论文之类的。我反而在 https://en.wikipedia.org/wiki/Virtualization 最下面的表中找到 OS-level virtualization 包含 docker

这是 Docker 的 [维基百科](https://en.wikipedia.org/wiki/Docker_(software)) 介绍

>  Docker provides an additional layer of abstraction and automation of operating-system-level **virtualization** on Windows and Linux.

Operating-system-level virtualization  的 [维基百科](https://en.wikipedia.org/wiki/Operating-system-level_virtualization)

> Operating-system-level virtualization, also known as containerization, refers to an operating system feature in which the kernel allows the existence of multiple isolated user-space instances.

我觉得文中使用『虚拟化（virtualization）技术』一词没有什么问题。

**Comment by jimms** [comment]
感謝分享，回報一下幾個錯字:
- 我们在这里还需要通过 **libcotainer** 提供的 **pivor_root** 或者 chroot => libcontainer 和 pivot_root

**Comment by michellewkx** [comment]
6666 学习了

**Comment by draveness** [comment]
> @jimms
> 感謝分享，回報一下幾個錯字:
> - 我们在这里还需要通过 **libcotainer** 提供的 **pivor_root** 或者 chroot => libcontainer 和 pivot_root 

好的，我改一下

**Comment by draveness** [comment]
> @michellewkx
> 6666 学习了

😃

**Comment by xiaoxubeii** [comment]
少有的能清晰的介绍docker底层原理的文章，足以证明作者的功底。

**Comment by V0idk** [comment]
条理清晰,深入浅出

**Comment by Bst2000** [comment]
UnionFS 那节的标题感觉有点准确，是不是改成 镜像管理 或者 存储驱动？如果我的建议不妥欢迎纠正。

**Comment by levicc** [comment]
pivor_root 或者 chroot这里不是很明白。chroot只是对一个进程生效，但是pivor_root，会unmount原来的root。如果用了它，那docker run之后，那不是影响了所有的进程了嘛

**Comment by yinwoods** [comment]
博主文章写的很赞！但是关于容器与镜像之间的关系，我有一个问题，当启动容器时，我们可以删除容器中原有的文件，而容器下层都是使用的镜像只读文件，那么这种删除的操作是如何实现的呢？

**Comment by cocowool** [comment]
"但是却有没有办法通过宿主机的网络与整个互联网相连"，这句可以修改一下。

**Comment by yanbingwei86** [comment]
读完还是有很多问题 比如：aufs或者overlay文件系统的镜像分层是怎么做的，每个docker容器的文件系统是共享的，还是单独mount了一个目录。等等

**Comment by Riven-Yang** [comment]
感谢！

**Comment by forezp** [comment]
牛逼

**Comment by lxkaka** [comment]
pretty cool

**Comment by fosteryang** [comment]
写的很用心，赞一个

**Comment by shengxinjing** [comment]
学习了很多，支持一下，请问博主的图使用什么软件画的 很好看

**Comment by depfsh** [comment]
难得有一篇文章把docker 说清楚

**Comment by iarchean** [comment]
文章帮助非常大，感谢！

**Comment by weihuacern** [comment]
感谢！获益良多！

**Comment by qianzhuoyao** [comment]
太棒了把

**Comment by jansona** [comment]
thx

**Comment by imageslr** [comment]
太赞了！

**Comment by zhanjixun** [comment]
如何搭建一个跟你一样的博客系统呢

**Comment by imageslr** [comment]
> 如何搭建一个跟你一样的博客系统呢

@zhanjixun 博客的开发框架应该是 jekyll，主题是 jasper
* jekyll https://github.com/jekyll/jekyll
* jasper https://jekyller.github.io/jasper/

博主有一篇相关博文：[jekyll 自动初始化 gitalk ](https://draveness.me/git-comments-initialize)。

希望对你有帮助。

**Comment by i0Ek3** [comment]
看作者的文章真的是一种享受，还看了作者的Go底层解析，是真的非常棒，对于自己走马观花式的学习深感羞愧。多向作者学习～

**Comment by ukinhappy** [comment]
https://mp.weixin.qq.com/s/QFtQ2Yy6YIRy__cJgx-Djg
博主可以看一下这篇文章

**Comment by draveness** [comment]
> https://mp.weixin.qq.com/s/QFtQ2Yy6YIRy__cJgx-Djg
> 博主可以看一下这篇文章

多谢，已经举报了...

**Comment by lamth** [comment]
many thank for the awesome post.

**Comment by draveness** [comment]
> many thank for the awesome post.

Thanks! But I'm quite curious about whether you read the original post or a translated version.

**Comment by wildwild** [comment]
从sohu转过来，谢大侠解释

**Comment by draveness** [comment]
> 从sohu转过来，谢大侠解释

SOHU 转过来，哪里啊

**Comment by yezihack** [comment]
文章采用的字体,好不习惯.还是因为我的浏览器问题?
![](https://upload-images.jianshu.io/upload_images/13827699-dd70260ad6ebd55a.png)

**Comment by draveness** [comment]
> 文章采用的字体,好不习惯.还是因为我的浏览器问题?

你浏览器的问题

**Comment by NominationP** [comment]
你好，我看完文章有几点想法，可否交流一下
1. 我之前对docker的了解就是可以用，且知道 namespace, control group, union filesystem 这三个linux中的技术构成了docker，以及三个技术的解决痛点
2. 但再问我底层我就知道，也不想知道了，我看了文章的题目，第一反应是我可以通过这篇文章更深层的了解这写底层，但是看完后我依然不知道
3. 我们写文章总是会看更多的文章，然后把认为重要的点，整合之后再写一遍，组成一篇新的文章。我现在是这样理解的，让我很困惑，就很难以理解有些评论：“难得有一篇文章把docker 说清楚”
4. 作者我们可以聊聊吗？吼吼吼

**Comment by draveness** [comment]
> 你好，我看完文章有几点想法，可否交流一下
> 
> 1. 我之前对docker的了解就是可以用，且知道 namespace, control group, union filesystem 这三个linux中的技术构成了docker，以及三个技术的解决痛点
> 2. 但再问我底层我就知道，也不想知道了，我看了文章的题目，第一反应是我可以通过这篇文章更深层的了解这写底层，但是看完后我依然不知道
> 3. 我们写文章总是会看更多的文章，然后把认为重要的点，整合之后再写一遍，组成一篇新的文章。我现在是这样理解的，让我很困惑，就很难以理解有些评论：“难得有一篇文章把docker 说清楚”
> 4. 作者我们可以聊聊吗？吼吼吼

我没有理解你的问题，你的问题是？

**Comment by llitfkitfk** [comment]
@NominationP 
你已经知道了
再底层就是010101了 没什么好知道的

**Comment by Crshi** [comment]
小瑕疵：但是却有没有办法 => 但是却又没有办法
博主真的厉害，向您看齐

**Comment by draveness** [comment]
> 小瑕疵：但是却有没有办法 => 但是却又没有办法
> 博主真的厉害，向您看齐

修复了

**Comment by Jed-Z** [comment]
文章质量很高，关注了～

**Comment by weihong363** [comment]
学习了，感谢博主

**Comment by letian0805** [comment]
> > @letian0805
> > 纠正下文章的开头，是“容器技术”，不是“虚拟化技术”
> 
> 有权威的出处么，我没有找到可信的资料，比如论文之类的。我反而在 https://en.wikipedia.org/wiki/Virtualization 最下面的表中找到 OS-level virtualization 包含 docker
> 
> 这是 Docker 的 [维基百科](https://en.wikipedia.org/wiki/Docker_(software)) 介绍
> 
> > Docker provides an additional layer of abstraction and automation of operating-system-level **virtualization** on Windows and Linux.
> 
> Operating-system-level virtualization 的 [维基百科](https://en.wikipedia.org/wiki/Operating-system-level_virtualization)
> 
> > Operating-system-level virtualization, also known as containerization, refers to an operating system feature in which the kernel allows the existence of multiple isolated user-space instances.
> 
> 我觉得文中使用『虚拟化（virtualization）技术』一词没有什么问题。

虚拟化包括：主机虚拟化、OS虚拟化。VirtualBox、qemu和vmware那种是主机虚拟化。在docker出现之前，我们说虚拟化等同于主机虚拟化。在docker出现后，我们不能将“虚拟化技术”等同“容器技术”。 虚拟化技术范围大于容器技术，docker也只是容器技术中的一种。
具体可以看维基百科https://en.wikipedia.org/wiki/Virtualization
![image](https://user-images.githubusercontent.com/14961048/95575707-be8dc680-0a61-11eb-872b-a4d120d9bd08.png)

@draveness

**Comment by WingLim** [comment]
[chroot_linux.go#L15](https://github.com/docker/docker-ce/blob/5f15cfcd9326e98ef2e08a7448eb316182fdc178/components/engine/pkg/chrootarchive/chroot_linux.go#L15)
Docker现在主要使用pivot_root来切换rootfs，当检测到当前进程位于user namespace的时候才会使用chroot。
使用pivot_root的主要原因之一是因为chroot容易逃逸。

**Comment by ancientmodern** [comment]
感谢作者，学到了很多。另外想请教下现在主流的overlay2存储驱动和本文中介绍的aufs存储驱动区别在哪里呢？

**Comment by draveness** [comment]
> @ancientmodern
> 感谢作者，学到了很多。另外想请教下现在主流的overlay2存储驱动和本文中介绍的aufs存储驱动区别在哪里呢？

速度会更快，具体区别可以搜一下，很容易的

**Comment by liyinxin** [comment]
感谢作者，非常棒的资料！！！！

**Comment by dwgeneral** [comment]
赞

**Comment by Donny1996** [comment]
虽然没有完全读懂，但是是真的写得好

**Comment by lr1848714606** [comment]
好文章,赞!

**Comment by MachaCroissant** [comment]
感谢分享！！

**Comment by azusachino** [comment]
谢谢分享

**Comment by Shellbye** [comment]
非常赞的总结

**Comment by cloudintheking** [comment]
非常棒

**Comment by xiaofengliao** [comment]
这文章实在是写的好，赞一下
